### PR TITLE
Local Library

### PR DESCRIPTION
**Pull request by coder1033**
### Endpoints
There are a total of 33 endpoints - 
- **Home**:
    ```
    Home page-
    URI: GET /catalog
    Status: 200
    Response: index.pug
    ```
- **Book**:
    ```
    List all books -
      URI: GET /catalog/books
      Status: 200
      Response: book_list.pug
    ```
    ```
    Each book detail -
      URI: GET /catalog/book/:id
      Status: 200
      Response: book_detail.pug
    ```
    ```
    Create Book -
      URI: GET /catalog/book/create
      Status: 200
      Response: book_form.pug

      URI: POST /catalog/book/create
      Request: { 
                    title: 'book title', 
                    author: 'book author',
                    summary: 'book summary', 
                    isbn: 'book ISBN no.', 
                    genre: 'book genre'
                }
      Status: 200
      Response: redirect(/catalog/book/:id)

    ```
    ```
    Update Book -
      URI: GET /catalog/book/:id/update
      Status: 200
      Response: book_form.pug

      URI: POST /catalog/book/:id/update
      Request: { 
                    title: 'book title', 
                    author: 'book author',
                    summary: 'book summary', 
                    isbn: 'book ISBN no.', 
                    genre: 'book genre'
                }
      Status: 200
      Response: redirect(/catalog/book/:id)

    ```
    ```
    Delete Book -
      URI: GET /catalog/book/:id/delete
      Status: 200
      Response: book_delete.pug

      URI: POST /catalog/book/:id/delete
      Request: { 
                    bookid: 'unique book id'
                }
      Status: 200
      Response: redirect(/catalog/books)

    ```

- **Author**:
    ```
    List all authors -
      URI: GET /catalog/authors
      Status: 200
      Response: author_list.pug
    ```
    ```
    Each author detail -
      URI: GET /catalog/author/:id
      Status: 200
      Response: author_detail.pug
    ```
    ```
    Create Author -
      URI: GET /catalog/author/create
      Status: 200
      Response: author_form.pug

      URI: POST /catalog/author/create
      Request: { 
                    first_name: 'first name of author', 
                    family_name: 'surname of author',
                    date_of_birth: 'birth date of author', 
                    date_of_death: 'death date of author'
                }
      Status: 200
      Response: redirect(/catalog/author/:id)

    ```
    ```
    Update Author -
      URI: GET /catalog/author/:id/update
      Status: 200
      Response: author_form.pug

      URI: POST /catalog/author/:id/update
      Request: { 
                    first_name: 'first name of author', 
                    family_name: 'surname of author',
                    date_of_birth: 'birth date of author', 
                    date_of_death: 'death date of author'
                }
      Status: 200
      Response: redirect(/catalog/author/:id)

    ```
    ```
    Delete Author -
      URI: GET /catalog/author/:id/delete
      Status: 200
      Response: author_delete.pug

      URI: POST /catalog/author/:id/delete
      Request: { 
                    authorid: 'unique author id'
                }
      Status: 200
      Response: redirect(/catalog/authors)

    ```
- **Genre**:
    ```
    List all genres -
      URI: GET /catalog/genres
      Status: 200
      Response: genre_list.pug
    ```
    ```
    Each genre detail -
      URI: GET /catalog/genre/:id
      Status: 200
      Response: genre_detail.pug
    ```
    ```
    Create Genre -
      URI: GET /catalog/genre/create
      Status: 200
      Response: genre_form.pug

      URI: POST /catalog/genre/create
      Request: { 
                  name: 'genre name'
                }
      Status: 200
      Response: redirect(/catalog/genre/:id)

    ```
    ```
    Update Genre -
      URI: GET /catalog/genre/:id/update
      Status: 200
      Response: genre_form.pug

      URI: POST /catalog/genre/:id/update
      Request: { 
                  name: 'genre name'
                }
      Status: 200
      Response: redirect(/catalog/genre/:id)

    ```
    ```
    Delete Genre -
      URI: GET /catalog/genre/:id/delete
      Status: 200
      Response: genre_delete.pug

      URI: POST /catalog/genre/:id/delete
      Request: { 
                    genreid: 'unique genre id'
                }
      Status: 200
      Response: redirect(/catalog/genres)

    ```
- **Book Instances**:
    ```
    List all book instances -
      URI: GET /catalog/bookinstances
      Status: 200
      Response: bookinstance_list.pug
    ```
    ```
    Each bookinstance detail -
      URI: GET /catalog/bookinstance/:id
      Status: 200
      Response: bookinstance_detail.pug
    ```
    ```
    Create Book Instance -
      URI: GET /catalog/bookinstance/create
      Status: 200
      Response: bookinstance_form.pug

      URI: POST /catalog/bookinstance/create
      Request: { 
                  book: 'Book object',
                  imprint: 'publisher',
                  status: 'availability status',
                  due_back: 'availability date'
                }
      Status: 200
      Response: redirect(/catalog/bookinstance/:id)

    ```
    ```
    Update Book Instance -
      URI: GET /catalog/bookinstance/:id/update
      Status: 200
      Response: bookinstance_form.pug

      URI: POST /catalog/bookinstance/:id/update
      Request: { 
                  book: 'Book object',
                  imprint: 'publisher',
                  status: 'availability status',
                  due_back: 'availability date'
                }
      Status: 200
      Response: redirect(/catalog/bookinstance/:id)

    ```
    ```
    Delete Book Instance -
      URI: GET /catalog/bookinstance/:id/delete
      Status: 200
      Response: bookinstance_delete.pug

      URI: POST /catalog/bookinstance/:id/delete
      Request: { 
                    bookinstanceid: 'unique bookinstance id'
                }
      Status: 200
      Response: redirect(/catalog/bookinstances)

    ```

### Database schema

Virtual fields in Book
```
url: String
```

Virtual fields in Genre
```
url: String
```

Virtual fields in Author
```
name
lifespan: String
url: String
date_of_birth_yyyy_mm_dd: DateTime
date_of_death_yyyy_mm_dd: DateTime
```

Virtual fields in Book Instance
```
url: String
due_back_formatted: DateTime
due_back_yyyy_mm_dd: DateTime
```

### Tests

**Automated test cases added**
- populatedb.js - It creates sample data on the above schema

**Manual test cases run**
- Checks all the endpoints
- Creates few test data

### CODE REVIEW

File: controllers/author-controller.js [hunk 1]
```diff
@@ -0,0 +1,213 @@
+const async = require("async");
```
Comment (GaurisankarJ) [inline controllers/author-controller.js:1]: Reorder please, NPM packages before local imports.

File: views/genre_form.pug [hunk 2]
```diff
@@ -0,0 +1,14 @@
+extends layout
```
Comment (GaurisankarJ) [inline views/genre_form.pug:1]: Is this file properly formatted?

File: controllers/author-controller.js (hunk 1 above)
Comment (coder1033) [reply controllers/author-controller.js:1]: So first there should be npm packages and then local imports. Right?

File: views/genre_form.pug (hunk 2 above)
Comment (coder1033) [reply views/genre_form.pug:1]: It's a .pug extension so prettier won't work here

**Comment by GaurisankarJ** [comment]
Please extract all required fields from the req object at the beginning of the function.
Eg:
```
const example = async (req, res, next) => {
  const { a, b } = req.body; // req.params or req.query
  ....,
  }
```

File: controllers/author-controller.js (hunk 1 above)
Comment (GaurisankarJ) [reply controllers/author-controller.js:1]: Yes.

File: views/genre_form.pug (hunk 2 above)
Comment (GaurisankarJ) [reply views/genre_form.pug:1]: Okay!

File: controllers/author-controller.js (hunk 1 above)
Comment (GaurisankarJ) [reply controllers/author-controller.js:1]: Try and keep them grouped and alphabetised.

Comment (coder1033) [reply controllers/author-controller.js:1]: Okay

**Review by GaurisankarJ — APPROVED** [review]

**Pull request merged**